### DISCUSSION TITLE
[RFC] Terminal UI

### DISCUSSION THREAD

**Discussion by anthonyshew** [vercel/turborepo#7802] [RFC]
# Thank you for feedback

The terminal UI has been released in 2.0! 🥳 

We appreciate everyone who participated and look forward to continuing to work with the Turbo community going forward.

**If you believe you've found a bug**, please report it in [Issues](https://github.com/vercel/turbo/issues).

**If you have an idea or enhancement you'd like to suggest**, please open a new [Discussion](https://github.com/vercel/turbo/discussions/categories/ideas).

---
## RFC
In Turborepo 1.13, we're releasing a new terminal UI that changes the way you experience your Turborepo locally.

Because this new experience is quite different from the previous one, we're creating this RFC to collect community feedback.

## What is the new terminal UI?

The new UI allows you to:
- **Toggle through the logs of each of your tasks with your arrow keys**, making them easier to follow and read
- **Interact with your persistent tasks** by entering and exiting the task’s shell, giving you the ability to re-run unit tests, for example

<details>
<summary>v1</summary>
![Screen capture of the new terminal UI](https://github.com/vercel/turbo/assets/35677084/01767bfe-f5f8-4d52-8faf-dc1439d99e05)
</details>

https://github.com/vercel/turbo/assets/35677084/25d711c8-cc43-4db0-9fe8-ab998d325001

## How do I enable the new terminal UI?

The new UI can be enabled two ways:
- Use the `--experimental-ui` flag when you run your `turbo` commands
- Add the `experimentalUI` key to your `turbo.json`

```json
{
  "experimentalUI": true
}
```

### How do I make a task interactive?

Set the `interactive` key to true on any task that you'd like to make interactive. Note that any `interactive` task must also have the `persistent` key set to `true`, as well.

```json
{
  "pipeline": {
    "test:dev": {
      "interactive": true,
      "persistent": true
    }
  }
}
```

## How do I give feedback?

We're excited to hear from you! Please comment below with any feedback you have, whether it be an issue or other general feedback. The core team will triage comments here and move issues to the Issues board as needed.

**Comment by weyert** [comment]
Thanks for the amazing work! It's looking great so far.

I have some feedback:
- If you use Warp.app it doesn't show the top bar as shown in the screenshot
- How do I scroll in task window? E.g. sometimes you might want scroll back 
- I think it would be nice to be able to ignore switching between cached tasks with suppressed logs (e.g. build tasks) see also: https://github.com/vercel/turbo/discussions/7779

**Reply by anthonyshew** [reply]
> If you use Warp.app it doesn't show the top bar as shown in the screenshot

Does making your window taller have the top bar show up? Was just messing around with a few things and noticed this myself.

**Reply by chris-olszewski** [reply]
> How do I scroll in task window? E.g. sometimes you might want scroll back

This hasn't been implemented yet, but is very doable. Thoughts on sensible keybindings to control the scroll? If we match existing conventions could do `Ctrl-P`/`Ctrl-N` or `k`/`j`.

**Reply by anthonyshew** [reply]
> could do Ctrl-P/Ctrl-N or k/j

I love my Vim binds as well but I'd think about keeping it more generalized. Ctrl-UpArrow, Ctrl-DownArrow?

**Reply by arlyon** [reply]
There are APIs for mouse scrolling which would be very cool also

**Reply by weyert** [reply]
Cool, yeah, think scrolling would be nice. 

Now while I thought of using a virtual term myself to get this all going I have to say I am impressed you were able to achieve it :) 
Looks like Rust got some nice crates related to it. Super cool. And I didn't know terminal supported mouse scrolling.

> Does making your window taller have the top bar show up? Was just messing around with a few things and noticed this myself.

Ah yeah, looks like making it a lot taller seems to work. Not sure if it is because it is trying to display all packages, which would be roughly ~50 packages (36 ^build tasks and then 10 `dev` persistent tasks)

**Reply by chris-olszewski** [reply]
First pass at scrolling: https://github.com/vercel/turbo/pull/7807
~~I'll can see how capturing mouse events works via `crossterm` later today as a scroll wheel would be very intuitive.~~
PR has been updated so mouse scroll wheel now work.

> Ctrl-UpArrow, Ctrl-DownArrow?

Ctrl-Up/Ctrl-Down aren't good candidates for this since macOS has those as default keybinds.

**Reply by anthonyshew** [reply]
Ah, heck, duh. I'd still wonder about something more intuitive for the non-Vimmers of the world.

**Reply by rafaeltab** [reply]
Any chance we could be able to customize the controls somehow? Maybe a `$HOME/.config/turbo/...` folder where we could add configuration for the current user.

I would love for there to be a programmable api that I could integrate into neovim myself. Perhaps an option to start a socket server, and run a command (Run the socket server and then pass the local address to the command as an argument). This could also be integrated with vscode with an extension.

If that is not an option I think actually the most optimal scenario would be a combination of the three.
- Vim motions for vimmers like myself, I'd prefer these keybinds to be as extensive as possible (`g`, `G`, `<C-d>`, `<C-u>`, `w`, `W`, etc).
- Arrows for the people who are not familiar with vim
- Scroll for people who don't like keyboard

**Comment by devjiwonchoi** [comment]
### Context

I am using `turbo` in a single Next.js app to benefit from remote cache.

### Problem

1. The build log for `next build` gets cleared. - Is this intended?
2. I need to manually enlarge the terminal size to view the rest of the logs.

### Info

Turbo: `1.13.0-canary.4`
 
```sh
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:32:11 PDT 2023; root:xnu-10002.41.9~7/RELEASE_ARM64_T6030
  Available memory (MB): 18432
  Available CPU cores: 11
Binaries:
  Node: 18.17.0
  npm: 9.6.7
  Yarn: N/A
  pnpm: 8.15.4
Relevant Packages:
  next: 14.2.0-canary.34 // Latest available version is detected (14.2.0-canary.34).
  eslint-config-next: 14.2.0-canary.28
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.2
Next.js Config:
  output: N/A
```

https://github.com/vercel/turbo/assets/120007119/d3c17eeb-0faf-401b-a36c-332f39a58ede

**Reply by chris-olszewski** [reply]
> The build log for next build gets cleared. - Is this intended?

This is a known behavior at the moment. The UI needs to be properly integrated with `--output-logs` to have the correct behavior when it comes to persisting logs.

> I need to manually enlarge the terminal size to view the rest of the logs.

The layout of the UI is currently hard coded and not designed well for small terminals. I'm working on making it more reactive to the terminal size.

**Reply by devjiwonchoi** [reply]
Thank you for answering!

**Reply by anthonyshew** [reply]
@devjiwonchoi We just put up a newer design on the latest canary. Would love for you to give it a try!

**Reply by devjiwonchoi** [reply]
Looks awesome!!

**Reply by weyert** [reply]
I like the newer design only wish the left column would take half the space. Maybe limit it to the maximum width of the package? Now it's 50/50

**Comment by imranbarbhuiya** [comment]
in my windows vscode terminal, pressing the arrow key moves 2 packages at a time, and pressing enter does nothing

https://github.com/vercel/turbo/assets/74945038/29fc34e2-8259-40cc-95da-cec7b7bdddad

also pressing ctrl+c keeps the logs in the same place 
![image](https://github.com/vercel/turbo/assets/74945038/66a8fddf-92f9-4ce6-8209-1b07c9c5c933)

**Reply by chris-olszewski** [reply]
Promoted this to an issue: https://github.com/vercel/turbo/issues/7808
The double key press issue should be addressed by https://github.com/vercel/turbo/pull/7809
Looking into why the terminal clear wouldn't trigger.

**Comment by CogitoBurrito** [comment]
How would it look like in CI/CD?

**Reply by chris-olszewski** [reply]
The UI is only usable in an interactive terminal so it is unusable in CI/CD. (Unless you do extra work to open up a psuedoterminal, but that wouldn't be recommended)

**Comment by imranbarbhuiya** [comment]
Once it becomes stable, will it replace the old UI? Although this UI looks good, most of the time I'll want to see all the logs at the same time instead of manually changing the package so keeping the existing UI and introducing a flag like `--interactive` to opt-in for the new UI will be great

**Reply by chris-olszewski** [reply]
Due to the fact that we'll always need to support running in non-interactive terminals, the old UI will remain. (Could either be flagged on or forced via something like `turbo build | cat`)

That being said, I'd love to try to address any shortcomings you have with the UI.

> I'll want to see all the logs at the same time instead of manually changing the package

A few things we've discussed:
 - Logs will be persisted to the terminal (space above the UI) on completion. Similar to how [`--log-order=grouped`](https://turbo.build/repo/docs/reference/command-line-reference/run#--log-order) works now.
 - Automatically jump to logs of a running task e.g. once `ui#build` finishes, jump to `web#build`. User can still use arrow keys to inspect finished tasks.

We're also open to any suggestions that would make the new UI an acceptable default.

**Reply by weyert** [reply]
> Automatically jump to logs of a running task e.g. once ui#build finishes, jump to web#build. User can still use arrow keys to inspect finished tasks.

I am not sure how this work? What happens when you need to build 30+ packages before `web#build`? Would cause a lot of jumping around, right?

**Reply by anthonyshew** [reply]
@weyert Just want to let you know. We tried it out and it didn't work out to be the experience we thought it was going to be. We all were excited to see it and then, once we saw it, quickly agreed it was definitely not right. :smile:

**Comment by steveluscher** [comment]
When an error is encountered, would you consider _leaving_ the experimental UI onscreen with the errored task made active? I was surprised to see the experimental UI close when there was an error.

**Reply by chris-olszewski** [reply]
Can you try `1.13.1-canary.0` which contains https://github.com/vercel/turbo/pull/7805 and see if that's the behavior you'd want?

**Reply by steveluscher** [reply]
Yeah, 100%. Thanks!

**Reply by anthonyshew** [reply]
We ironed out a few more things on the most recent canary, as well, including a redesign. Would love for you to take another look!

**Comment by dBianchii** [comment]
I have two feedbacks:

1. I didn't see any reason to use the new experimental UI on linter/formatter and typecheck tasks. I tried using it, but it was very hard to see what was failing, and it immediatelly kicked me out of the UI and I wasn't able to see where the errors were happening. I guess this UI is more useful for long running tasks like running servers with dev scripts or drizzle/prisma studio.

2. I can't Cntrl+click links that appear in the terminal:
![image](https://github.com/vercel/turbo/assets/17372417/39f806bc-c3f7-4c08-adad-e6f19aa2a4c4)

**Reply by chris-olszewski** [reply]
> I tried using it, but it was very hard to see what was failing

Can you try the latest canary? It now persists all task logs to the terminal so after `turbo` exits you can see failures/outputs.

> I can't Cntrl+click links that appear in the terminal

Looking into what it'll take to support this.

**Comment by haydnhkim** [comment]
I think it should take the opposite approach and consider providing a simple terminal ui option.

- inherited stdin option
- option to turn off task names on stdout and output to raw (mixed log is fine)

The new terminal UI feels like it limits a lot of the functionality of the original terminal app.
It also feels less responsive.
I would like to be able to view and copy the output logs raw.

**Reply by chris-olszewski** [reply]
> inherited stdin option

This is deemed as too dangerous and unintuitive to support. Dangerous as `stdin` is an uncaptured input so users could poison the shared cache. Unintuitive because inheriting stdin means users wouldn't be able to send specific input to specific tasks, just forward input to all tasks that inherit stdin.

> option to turn off task names on stdout and output to raw (mixed log is fine)

This is achievable right now with the [`--log-prefix=none`](https://turbo.build/repo/docs/reference/command-line-reference/run#--log-prefix) flag

> It also feels less responsive.

Can you elaborate on this? What part of the UI or invocation of `turbo` feels less responsive? Is it task output?

**Reply by haydnhkim** [reply]
@chris-olszewski 
Thanks for pointing me to the `--log-prefix` setting.

`stdin` forwarding is primarily needed when requesting specific functionality from a running process in the `dev` pipeline.
For example)
https://github.com/vercel/turbo/issues/1235
https://github.com/vercel/turbo/issues/7332

I want to pass a keyword like `rs` or `o` regardless of the cache.
How about setting certain in-pipeline options differently from the experimentalUI options?

```
"pipeline": {
  "dev": {
    "stdin": true,
  }
}
```

Additionally, Maybe it would be nice to allow experimentalUI options to be set in the pipeline as well.
Perhaps there is a complexity to making it applicable only to the top-level pipeline.

---

I feel that when running with experimentalUI, the speed at which text is entered and logs are output is slower than in normal mode.
This is probably due to the fact that I'm using an Intel Mac.
I'm using Warp.app and it doesn't seem to expose all of the UI as weyert mentioned.

**Comment by juliusmarminge** [comment]
When everything is done, it'd be nice to be able to look at the output before "exiting". Right now it just exits as soon as the last task is done so you can't see any of the outputs.

**Reply by anthonyshew** [reply]
Logs should be being persisted to your terminal with latest releases!

**Comment by yamcodes** [comment]
The new UI looks great, but it feels like you need to reveal too much of the terminal for more than one task to be visible in the switcher. In other words, the terminal height needs to be set too high for the switcher to show multiple tasks. Normally I would have my terminal reveal maybe 10-15 lines, but to see multiple tasks I need 30-35.

What would be nice to have is a way to resize the interactive portion of the experimental UI so that we can select our own sweetspot between interactive terminal space and task switcher height.

**Reply by AndreaPontrandolfo** [reply]
Yes agree, the fact that the buffer size doesn't auto-rescale is unbearable

**Reply by anthonyshew** [reply]
Heard! There's a new design that should meet this need on the most recent canary. Would love for you to take a look!

**Comment by MickL** [comment]
What bothers me is that when I only have on task running (which is common e.g. for Drizzle) it still shows this task UI and I need to jump into the task with enter key. Instead Turbo could notice I am executing just one task and just show me the output of this task with nothing around it.

**Comment by MickL** [comment]
For me the output is cut off so I cant see what use action I am doing

<img width="806" alt="Bildschirmfoto 2024-04-03 um 21 58 13" src="https://github.com/vercel/turbo/assets/5103210/b24dff06-dd51-4d55-b175-88e23dceb10f">

**Reply by anthonyshew** [reply]
We've got a new design on the most recent canary that should hopefully help with this. Give it a look!

**Comment by juliusmarminge** [comment]
![CleanShot 2024-04-03 at 23 38 01@2x](https://github.com/vercel/turbo/assets/51714798/2f5e7412-e5db-4843-a1c8-b324e3f2ce42)

indentation gets weird sometimes...

**Reply by anthonyshew** [reply]
Try out the latest canary! Should be fixed.

**Comment by juliusmarminge** [comment]
file-links aren't clickable:
https://github.com/vercel/turbo/assets/51714798/b44a877d-8d4a-43bd-a855-f1c32668ee6b

**Comment by Willem-Jaap** [comment]
Thanks for this!
The terminal looks great but takes up a lot of screen real estate.

I think there is some space that can be freed up, maybe a `'experimentalUIStyle: "compact" | "default"' prop?

![image](https://github.com/vercel/turbo/assets/67187467/af6f1c63-fde3-4ca1-bb28-737aa0cc8392)

Maybe the interactions can be moved to top right and using symbols.
`↑↓ to navigate, ⏎ to interact, ctrl+z to stop`

**Reply by anthonyshew** [reply]
We did a redesign that should help with this! Try out the latest canary. 👍

**Comment by blankdreams2** [comment]
![image](https://github.com/vercel/turbo/assets/94078395/9b704c92-eba8-4e09-83cf-c066cab0a233)

It's working great for my project after some more testing. Do agree that it would be great if you provided different layouts such as compact, full, or only the necessary.

The scroll is smooth, `but` it's slow IMO. If I have a long error or log messages then it might hinder some time in scrolling. And the feel of having another terminal inside your terminal 🤣. Overall a great addition

**Comment by juliusmarminge** [comment]
Perhaps not an issue in turbo, but i can't enter into a task and interact with it using Warp. Even if i press Enter, i still just navigate between the tasks and not up and down inside the task

https://github.com/vercel/turbo/assets/51714798/cda2bc11-e698-41f2-908f-cbeab91d4fe2

**Comment by juliusmarminge** [comment]
Is it necessary for all logs to be prefixed? Maybe the header should be bold and colorful instead of the line inside of the block?

![CleanShot 2024-04-07 at 22 45 49@2x](https://github.com/vercel/turbo/assets/51714798/aa466081-df76-4770-a9c3-9f94f2f4adc3)

**Reply by juliusmarminge** [reply]
Ok with a bit more info in the task it looks a bit better, still think that the header should be more visible and remove the prefix inside of the block?

![CleanShot 2024-04-07 at 22 48 41@2x](https://github.com/vercel/turbo/assets/51714798/01300951-4468-46ce-bfec-0d29829fd33c)

**Reply by anthonyshew** [reply]
We did some redesigning that should address these. Try the latest canary!

**Comment by ColinCee** [comment]
Is the interaction ability bugged? Pressing enter doesn't allow me to send commands to jest watch. a, f, o, q and nothing happens

on Turbo v1.13.2

https://github.com/vercel/turbo/assets/11246116/c5c155cc-94e1-4364-8992-5464b9453b3b

my task config

```
    "test:watch": {
      "dependsOn": ["@right-track/common#build"],
      "cache": false
    },
```

I'm using zsh in vscode integrated terminal

**Reply by anthonyshew** [reply]
You'll need to mark the task as interactive with `"interactive": true`.

**Reply by ColinCee** [reply]
Interesting, so by default it's not interactive? The message with "Press 'Enter' to interact..." is misleading in that case

**Reply by anthonyshew** [reply]
Fair! cc @chris-olszewski

**Reply by MickL** [reply]
I didnt set this option but it was interactive for me on macOS

**Reply by ColinCee** [reply]
Adding `"interactive": true` to the task definition fixed the issue, not sure why you don't need it @MickL but it doesn't work for me without it

**Reply by anthonyshew** [reply]
We've updated this behavior. Any `persistent` task is now interactive by default!

**Comment by pjmlp** [comment]
While the UI looks nice, it is completly unusable for me, I am completly lost on how to find errors, and documentation doesn't help.

Had to disable it.

**Reply by anthonyshew** [reply]
This is currently an experimental feature so we haven't documented it yet. Could you be more specific about the problems you had?

**Reply by pjmlp** [reply]
Basically I was left with "Build error" kind of message, without being able to see the actual compilation trace that lead to the error.

I had to revert to the old UI, to finally being able to read that environment variables validation was giving an error regarding current configuration, something that with a plain "Build Error", I would never find out.

**Reply by anthonyshew** [reply]
It sounds like we hadn't implemented scrolling yet, which should be available now. If you can, try out the latest canary. Interested to hear what you think.

**Comment by stuible** [comment]
Loving this new UI so far, appreciate all the work that's gone into it!  One request, it would be useful to be able to switch to a view that combines the all the logs into one stream (without having to exit, disable the new UI and restart the application).

**Reply by anthonyshew** [reply]
This was something that we thought about - but, unfortunately, can't support due to how some CLIs paint to the terminal. A streamed log can't be interactive, some CLIs hard overwrite liners with their spinners, and a litany of other issues. This new UI is meant to solve many of those problems (along with creating an overall better experience) in one swoop.

If you do still like the log stream, you'll still be able to set it with [`--log-order=stream`](https://turbo.build/repo/docs/reference/command-line-reference/run#--log-order).

**Comment by samcx** [comment]
There is a flash of the UI when you get a full turbo.

**Comment by Willem-Jaap** [comment]
I just tried the latest canary, the layout looks more organized (which resolves my previous comment https://github.com/vercel/turbo/discussions/7802#discussioncomment-9008691), great job!

However, the tasks show a loading spinner for long running tasks like dev which is quite distracting
![image](https://github.com/vercel/turbo/assets/67187467/17ef61ad-fcb6-4e76-936a-2d6d044a971a)

**Comment by tungtranpendle** [comment]
Hi @anthonyshew, the new UI is great but may I ask how do I copy from the raw output of the latest experimental UI? It seems not compatible with the default terminal, hope you will support that feature

**Comment by dBianchii** [comment]
I can't highlight, or copy from the terminal.

**Comment by trungtin** [comment]
Note here if someone using iterm2 like me and cannot make it scroll.

Go to Profile > Terminal > Enable mouse reporting:
- Report mouse wheel events: enable to make the panel scrollable
- Report mouse clicks & drags: disable to be able to highlight and copy from terminal (or temporary disable by holding alt/option key)

**Comment by mstade** [comment]
This new UI is FIRE 🔥 Thank you! 🙏

**Comment by weyert** [comment]
How do we disable the UI in the latest canary build?

**Reply by venables** [reply]
set `"ui": "stream"` in your `turbo.json`

https://turbo.build/repo/docs/reference/configuration#ui

**Reply by ehaynes99** [reply]
Is there a way to do it on an individual basis, such as an env var? I don't want to disable it for my whole org, as some might like it, but it's not for me.

**Comment by MickL** [comment]
Feedback for `1.13.4-canary.5`: 

- Scrolling with trackpad doesnt seem to work. I scroll down and it scrolls up again and again.
- When starting just one task I cant enter interact with it. I can scroll (more or less...) but I when I press enter it doesnt work with confirming Drizzle Studio.
- Even when disabling the new UI I cant interact with the task AND it is cut off as well. Same for 1.13.3: <img width="811" alt="Bildschirmfoto 2024-06-03 um 11 25 45" src="https://github.com/vercel/turbo/assets/5103210/4aed22a8-47a5-4045-9de7-1dbb53a36916">
- When starting just one task (e.g. with `--filter="..."`) there should be no left sidebar. Actually in this case the whole terminal UI can be deactivated
- Otherwise when starting just one task and there is a left sidebar, the task should be interactive immediately. It is very scary to do something like `drizzle:push` and then it asks "Do you really want to push changes? > Yes, No " and I need to press enter to get into the task
- When I scroll in the terminal it scrolls the tasks on the left but not visible logs. This is very annoying because I need to go in the task first to be able to scroll.

Are you guys even reading this? I am afraid you will release the new version of Turbo with the UI that is not usable at all atm (see point 1-3).

**Comment by jamesvclements** [comment]
Looks like some items are hidden on the left unless I manually press down to reveal them:

https://github.com/vercel/turbo/assets/20052710/41d91e08-b84c-4706-bd1c-26c246818c4f

**Reply by anthonyshew** [reply]
Appreciate this report. I've converted it to an issue as #8310. Let's follow up there.

**Comment by jamesvclements** [comment]
It also seems like native inertia scrolling isn't supported, so scrolling through feels quite slow / sticky compared to the normal terminal, on a Mac OS trackpad / magic mouse at least

**Reply by ehaynes99** [reply]
I don't think that's possible to solve that in a TUI. Terminals don't have scrolling; it's a feature of a terminal _emulator_, and that's a desktop application that can take advantage of OS features. A TUI is a terminal application that effectively has no scrollback,

**Comment by siawyoung** [comment]
This new UI breaks when I'm running it in husky precommit:

In my root package.json:

```
    "lint": "turbo lint",
```

and in `.husky/pre-commit`:

```
pnpm lint
```

when running this, it gives:

```
$ git commit -m "Test commit"
• Packages in scope: ....
• Running lint in 8 packages
• Remote caching disabled
  × internal errors encountered: receiver dropped,receiver dropped,receiver dropped
```

The only way to fix this that I've found dso far is to add `"ui": "stream"` to the root `turbo.json`.

**Reply by anthonyshew** [reply]
Thanks for this report. I've created a new issue from this at #8311. We will follow up there.
